### ISSUE TITLE
unexpected initialization order of settings overridden in source dependencies

### ISSUE THREAD

**Issue by jastice** [sbt/sbt#3850]
Overriding settings of a source dependency build seems to happen in the wrong order.
When a source dependency explicitly defines a setting, it is not possible to override it from the build.sbt of the project consuming the dependency.

## steps

Clone project: https://github.com/jastice/buildref1
This project contains the main project `buildref1` which depends on another project `buildref2`.
`buildref2` defines a value for the setting `ideaDownloadDirectory` which is provided by a plugin.
`buildref1` attempts to override this value in it's build.sbt

## problem

```
sbt:parentBuild> {file:/Users/jast/playspace/buildref1/buildref2/}/ideaDownloadDirectory
[info] set in buildref2
sbt:parentBuild> {file:/Users/jast/playspace/buildref1/buildref2/}otherRoot/ideaDownloadDirectory
[info] set in buildref2
```

## expectation

```
sbt:parentBuild> {file:/Users/jast/playspace/buildref1/buildref2/}/ideaDownloadDirectory
[info] set in buildref1 for otherBuild
sbt:parentBuild> {file:/Users/jast/playspace/buildref1/buildref2/}otherRoot/ideaDownloadDirectory
[info]set in buildref1 for otherRoot
```

## notes

sbt version: 1.0.4

**Comment by jastice** [comment]
Workaround for the moment: define `onLoad in Global` to modify the settings after loading. It requires a flag to avoid an infinite recursion because modifying settings will re-trigger onLoad and will recalculate settings, which makes the whole load a little slower, so it's a rather unsatisfactory workaround.

Looks a bit like this:
```scala
onLoad in Global := { state: State =>
  val ex = Project.extract(state)
  val settingsAppliedKey = SettingKey[Boolean]("newSettingsApplied")
  val settingsAreApplied = ex.getOpt(settingsAppliedKey in Global).getOrElse(false)
  if (settingsAreApplied) state // need this to avoid infinite recursive apply
  else {
    val modifiedSettings =
      applyNewSettings(buildRef2) :+ (settingsAppliedKey in Global := true)
    ex.append(modifiedSettings, state)
  }
}.compose((onLoad in Global).value)
```

**Comment by jastice** [comment]
Furthermore, changing settings in `onLoad` can break the `apply` command.
For example, using a command sequence like this:
`;apply -cp someSettings.jar AddSomeCommand; someCommand`
will apply the settings from `AddSomeCommand`, which causes `onLoad` to trigger again, which will in turn effectively *remove* the new settings, thus `someCommand` will fail.

Not sure if this is bug or as expected? Not as I would expect, anyway.

**Comment by dwijnand** [comment]
Yeah I was talking to a colleague about how the API around `Extracted#append` and `SessionSettings` is confusing, unclear and under-documented.. He was hitting effectively the same issue around using `+` and `Extracted#append`.

**Comment by jastice** [comment]
Any workaround to this? I tried to do it as in sbt-structure and use `appendRaw`:
```
val appended = ex.session.appendRaw(modifiedSettings)
reapply(appended, state)
```

This doesn't break the other commands added by sbt-structure, but it also doesn't seem to modify the source dependency's settings correctly. So basically I am out of ideas how to modify the dependency build and still have things work.

**Comment by dwijnand** [comment]
I've not spent any time trying to figure out how to make everything as expected.

**Comment by 2m** [comment]
Interesting. I tried `appendRaw` and `reapply` and it worked for my use case which is a command that overrides a setting. So when cross-running that command the `scalaVersion` was not reset (contrary to when I was using `Extracted#append`).

I also found that with `appendRaw` I was able to override a setting from `Global` scope but not from `ThisBuild`. And it was another way around when using `Extracted#append`: able to override a setting from `ThisBuild`, but not from `Global`.

**Comment by dwijnand** [comment]
> I also found that with `appendRaw` I was able to override a setting from `Global` scope but not from `ThisBuild`. And it was another way around when using `Extracted#append`: able to override a setting from `ThisBuild`, but not from `Global`.

![fuck this shit gif](http://i0.kym-cdn.com/photos/images/original/000/385/009/23e.gif)

**Comment by jvican** [comment]
FTR, I've also stumbled upon this, and I solved it my way. In my opinion, source dependencies are fundamentally broken until this ticket is fixed. This is probably my biggest gripe with sbt at the moment.
  
I worked around this issue one month ago in bloop: https://github.com/scalacenter/bloop/blob/master/project/BuildPlugin.scala#L165-L195. It handles session settings correctly, and it also fixes the implementation of plugin cross which is at the moment of writing is broken. I hijacked the original implementation in https://github.com/scalacenter/bloop/blob/master/project/BuildPlugin.scala#L197-L200 where `WorkingPluginCross` is at https://github.com/scalacenter/bloop/blob/master/project/WorkingPluginCross.scala.

**Comment by jvican** [comment]
For those hitting on this issue too, I recommend that `set` is used instead of the state API directly. I think this solution is simpler than interfacing with the state API directly. If someone just wants to change the scala version of a sourcedep, something like this would work:
  
```scala
onLoad in Global := { state: State =>
  """set scalaVersion in mySourceDep := "2.12.4"""" :: state
}.compose((onLoad in Global).value)
```
  
Remember that `set` can redefine several settings and tasks in only one command by wrapping what follows set with a block whose final expression returns a setting or list of settings.

**Comment by jastice** [comment]
Suggestion from @eed3si9n: if we control the dependency project, we can check for existence of potentially overridden settings, and set them only if not available. @mutcianm

**Comment by mutcianm** [comment]
Another suggestion from @eed3si9n to extract settings of which we want to have initialization control into source based sbt plugins inside a project and define explicit dependencies via `trigger` and `requires`
